crossdock/server/server.py

```diff
@@ -77,25 +84,55 @@ def head(self):
 def make_app(server):
     return tornado.web.Application(
         [
-            (r"/", MainHandler),
-            (r"/start_trace", MainHandler, (dict(server=server, method=Server.start_trace))),
-            (r"/join_trace", MainHandler, (dict(server=server, method=Server.join_trace))),
+            (r'/', MainHandler),
+            (r'/start_trace', MainHandler, (dict(server=server, method=Server.start_trace))),
+            (r'/join_trace', MainHandler, (dict(server=server, method=Server.join_trace))),
         ], debug=True)
 
 
 # TChannel handlers
 def tchannelNotImplResponse():
-    return TraceResponse(notImplementedError="python <=> tchannel not implemented")
+    return TraceResponse(notImplementedError='python <=> tchannel not implemented')
+
 
+def make_tchannel(port):
+    global tchannel
+    tchannel = TChannel('python', hostport='localhost:%d' % port, trace=True)
 
-@tchannel.thrift.register(service.TracedService)
-def startTrace(request):
-    return trace_response_to_thriftrw(service, tchannelNotImplResponse())
+    service = get_thrift_service(service_name='python')
 
+    @tchannel.thrift.register(service.TracedService, method='joinTrace')
+    @tornado.gen.coroutine
+    def join_trace(request):
+        jtr = request.body.request if request.body.request else None
+        res = yield handle_trace_request(jtr)
+        raise tornado.gen.Return(res)
 
-@tchannel.thrift.register(service.TracedService)
-def joinTrace(request):
-    return trace_response_to_thriftrw(service, tchannelNotImplResponse())
+    @tornado.gen.coroutine
+    def handle_trace_request(trace_request):
+        span = get_current_span()
+        observed_span = service.ObservedSpan(
+            traceId='%x' % span.trace_id,
+            sampled=span.is_sampled(),
+            baggage=span.get_baggage_item(constants.baggage_key)
+        )
+
+        trace_response = TraceResponse(span=observed_span, notImplementedError='')
+
+        if trace_request and trace_request.downstream is not None:
+            downstream_trace_resp = yield call_downstream(span, trace_request.downstream)
+            observed_span = service.ObservedSpan(
+                traceId=downstream_trace_resp.span.traceId,
+                sampled=downstream_trace_resp.span.sampled,
+                baggage=downstream_trace_resp.span.baggage
+            )
+            downstream_trace_resp = TraceResponse(span=observed_span, notImplementedError='')
+            downstream_trace_resp = trace_response_to_thriftrw(service, downstream_trace_resp)
```

Comment:
this function again looks like it modifies its arguments, can it not simply return TraceResponse?